Improve documentation on the consumer side (#28)
osmosis-labs/mesh-security: docs/consumer/VirtualStaking.md

> stakers have delegated at a ration of 2:3. This gets worse if eg. someone unbonds all A. We end up with 120 requested for B,

Reviewer: ```suggestion
stakers have delegated at a ratio of 2:3. This gets worse if eg. someone unbonds all A. We end up with 120 requested for B,
```

but 80 delegated to A and 20 delegated to B.

To properly handle this, we should use Redelegations to rebalance when the amounts are updated. 
This may be quite some messages (if we have 50 validators, each time we stake one more, we need to 
add a bit to that one and decrease the other 50). We describe a solution to this in the next section.

### Epochs

> For efficiency rewards will be withdrawn for all cross-stakers at a regular rhythm, once per epoch.

Reviewer: ```suggestion
For efficiency, rewards will be withdrawn for all cross-stakers at a regular rhythm, once per epoch.
```

This is done to reduce the number of IBC messages sent, and reduce computation (gas) overhead.

> Note that one provider will like have to withdraw for dozens of validators and then distribute

Reviewer: ```suggestion
Note that one provider will likely have to withdraw from dozens of validators and then distribute
```
?

to the proper cross-stakers, so this is a significant amount of computation, and shouldn't be
triggered every few blocks.

While staking and unstaking can be quite cheap below the max cap, it becomes quite expensive once
the number of tokens exceeds the max cap, as it will require a rebalance. If a provider is
over the max cap and someone stakes 10 tokens on validator V, the provider will have to unstake
some tokens proportionally from every other validator. Likewise, if someone unstakes 10 tokens,
it will trigger a staking on the remaining validators to keep the total at max cap.

> It makes sense to do this rebalance only once per epoch, as it is quite expensive. These could

Reviewer: Add a "CosmWasm Contract" sub-title here?

be different epoch periods for two different purposes, but it is recommended to keep them
the same for simplicity. However, we do suggest that the epochs of each provider be offset,

> so we don't blog up the network with a bunch of IBC messages at the same time.

Reviewer: ```suggestion
so we don't clog up the network with a bunch of IBC messages at the same time.
```

Implementation of this will require another binding with the native Go module. Our proposed
mechanism is to add a special `SudoMsg` that is triggered once per epoch on each contract with
a max cap. Each contract will have a `next_epoch` value and they all will share an `epoch_length`
parameter.

Calling `sudo` will be done in EndBlock once per epoch, and will have a gas limit to prevent
any freezing of the chain. We suggest something like 50-100% of the normal block gas limit.
The calling code should also catch any error (including out of gas) and log it, but not fail.
It will revert all state changes made by the `sudo` message, mark it as ran and continue.
This provides a way to trigger the contract regularly with a large gas allowance, but not
be susceptible to any DoS attacks.

```rust
#[cw_serde]
pub enum SudoMsg {
  Rebalance{},
}
```

> ### Module

Reviewer: ```suggestion
### SDK Module
```

```go
type StakePermission struct {

>   // Limit we cap the virtual stake of this converter.
>   // Defined as a number of "virtual native tokens" this can mint.

Reviewer: ```suggestion
  // Limits the cap of the virtual stake of this converter.
  // Defined as a number of "virtual native tokens" this module can mint.
```
?

  MaxStakingRatio: sdk.Int,
  // Next time (unix seconds) we trigger the sudo message
  // When the contract is first registered, it is set to block time + epoch length
  NextEpoch: uint64,
}
```

Beyond MVP, we wish to add the following functionality:

* Provide configuration for optional governance multiplier (eg 1 virtual stake leads to 1 tendermint power, but may be 0 or 1 or even 0.5 gov voting power)

> ### Reward Withdrawls

Reviewer: ```suggestion
### Reward Withdrawals
```

The Virtual Staking Module uses an [EndBlock hook called once per epoch](#epochs)
(param setting, eg 1 day).
This will trigger a `SudoMsg` to each Virtual Staking contract, allowing them to both do a rebalancing
with any staking/unstaking in that period, as well as withdraw rewards from all validators.

> When the epoch finishes for one Converter, the Module will withdraw rewards from all delegations

Reviewer: ```suggestion
When the epoch finishes for one Converter, the Virtual Staking Module will withdraw rewards from all delegations
```
